client-assertion-validation/src/test/scala/it/pagopa/interop/clientassertionvalidation/ClientAssertionValidationSpec.scala

```diff
@@ -0,0 +1,229 @@
+
+      validateClientAssertion(clientId.toString.some, assertion, clientAssertionType, grantType)(
+        jwtValidator
+      ) shouldBe Left(NonEmptyList.of(JtiNotFound))
+    }
+
+    "fail when issuer is missing" in {
+      val assertion = fastClientAssertionJWT(issuer = None)
+
+      validateClientAssertion(clientId.toString.some, assertion, clientAssertionType, grantType)(
+        jwtValidator
+      ) shouldBe Left(NonEmptyList.of(IssuerNotFound))
+    }
+
+    "fail when IAT is missing" in {
+      val assertion = fastClientAssertionJWT(iat = None)
+
+      validateClientAssertion(clientId.toString.some, assertion, clientAssertionType, grantType)(
+        jwtValidator
+      ) shouldBe Left(NonEmptyList.of(IssuedAtNotFound))
+    }
+
+    "fail when EXP is missing" in {
+      val assertion = fastClientAssertionJWT(expirationTime = None)
+
+      validateClientAssertion(clientId.toString.some, assertion, clientAssertionType, grantType)(
+        jwtValidator
+      ) shouldBe Left(NonEmptyList.of(ExpirationNotFound))
+    }
+
+    "fail on algorithm not allowed" in {
+      val ecKey = new ECKeyGenerator(Curve.P_256).generate
+      val ecKid = ecKey.computeThumbprint().toJSONString
+
+      val assertion =
+        fastClientAssertionJWT(algorithm = "ES256".some, kid = ecKid.some, privateKeyPEM = ecKey.toJSONString.some)
+
+      validateClientAssertion(clientId.toString.some, assertion, clientAssertionType, grantType)(
+        jwtValidator
+      ) shouldBe Left(NonEmptyList.of(AlgorithmNotAllowed("ES256")))
+    }
+
+    "fail when digest algorithm is not allowed" in {
+      val digest = Digest(alg = "invalid", value = "2cf24dba5fb0a30e26e83b2ac5b9e29e1b161e5c1fa7425e73043362938b9824")
+      val assertion = fastClientAssertionJWT(customClaims = Map(DIGEST_CLAIM -> digest))
+
+      validateClientAssertion(clientId.toString.some, assertion, clientAssertionType, grantType)(
+        jwtValidator
+      ) shouldBe Left(NonEmptyList.of(InvalidHashAlgorithm))
+    }
+
+    "fail when digest hash value length is not valid" in {
+      val digest    = Digest(alg = "SHA256", value = "tooshort")
+      val assertion = fastClientAssertionJWT(customClaims = Map(DIGEST_CLAIM -> digest))
+
+      validateClientAssertion(clientId.toString.some, assertion, clientAssertionType, grantType)(
+        jwtValidator
+      ) shouldBe Left(NonEmptyList.of(InvalidHashLength("SHA256")))
+    }
+
+    "fail when digest algorithm is missing" in {
+      val digest    = Map("value" -> "2cf24dba5fb0a30e26e83b2ac5b9e29e1b161e5c1fa7425e73043362938b9824")
+      val assertion = fastClientAssertionJWT(customClaims = Map(DIGEST_CLAIM -> digest))
+
+      validateClientAssertion(clientId.toString.some, assertion, clientAssertionType, grantType)(
+        jwtValidator
+      ) shouldBe Left(NonEmptyList.of(DigestClaimNotFound("alg")))
+    }
+
+    // TODO This validation returns DigestClaimNotFound(alg) instead of DigestClaimNotFound(value)
```

Comment:
Is this still valid?

Reply:
this is one point I wanted to discuss today.
The code seems correct, the expected error type is raised, but the claim in the error is not 🤔